src/predicate.rs

```diff
@@ -82,20 +114,104 @@ impl<'a> Predicate for Attr<&'a str, &'a str> {
     }
 }
 
-impl<'a> Predicate for Attr<&'a str, ()> {
+/// Matches Element Node containing attribute `N` which matches function `F`.
+///
+/// There is no equivalent CSS selector.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub struct AttrMatches<N, V>(pub N, pub V);
+
+impl<'a, F: Fn(&str) -> bool> Predicate for AttrMatches<&'a str, F> {
+    fn matches(&self, node: &Node) -> bool {
+        node.attr(self.0).map_or(false, &self.1)
+    }
+}
+
+/// Matches Element Node containing attribute `N` which contains pattern `V`.
+///
+/// This is equivalent to the `[attr*=val]` CSS selector.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub struct AttrContains<N, V>(pub N, pub V);
+
+impl<'a> Predicate for AttrContains<&'a str, &'a str> {
+    fn matches(&self, node: &Node) -> bool {
+        node.attr(self.0).map_or(false, |a| a.contains(self.1))
+    }
+}
+
+/// Matches Element Node containing attribute `N` which starts with pattern `V`.
+///
+/// This is equivalent to the `[attr^=val]` CSS selector.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub struct AttrStartsWith<N, V>(pub N, pub V);
+
+impl<'a> Predicate for AttrStartsWith<&'a str, &'a str> {
+    fn matches(&self, node: &Node) -> bool {
+        node.attr(self.0).map_or(false, |a| a.starts_with(self.1))
+    }
+}
+
+/// Matches Element Node containing attribute `N` which ends with pattern `V`.
+///
+/// This is equivalent to the `[attr$=val]` CSS selector.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub struct AttrEndsWith<N, V>(pub N, pub V);
+
+impl<'a> Predicate for AttrEndsWith<&'a str, &'a str> {
+    fn matches(&self, node: &Node) -> bool {
+        node.attr(self.0).map_or(false, |a| a.ends_with(self.1))
+    }
+}
+
+/// Matches Element Node containing attribute `N` contains the word `V` surrounded by either the
+/// edges of the string, or whitespace.
+///
+/// This is equivalent to the `[attr~=val]` CSS selector.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub struct AttrContainsWord<N, V>(pub N, pub V);
+
+impl<'a> Predicate for AttrContainsWord<&'a str, &'a str> {
+    fn matches(&self, node: &Node) -> bool {
+        node.attr(self.0).map_or(false, |a| a.split_whitespace().any(|w| w == self.1))
+    }
+}
+
+/// Matches Element Node containing attribute `N` being either exactly the value `V` or `V`
+/// followed by `"-"`, then some string.
+///
+/// This is equivalent to the `[attr|=val]` CSS selector.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub struct AttrLang<N, V>(pub N, pub V);
+
+impl<'a> Predicate for AttrLang<&'a str, &'a str> {
+    fn matches(&self, node: &Node) -> bool {
+        node.attr(self.0).and_then(|a| a.split('-').next()).map_or(false, |a| a == self.1)
+    }
+}
+
+/// Matches Element Node containing attribute `N`.
+///
+/// This is equivalent to the `[attr]` CSS selector.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub struct HasAttr<N>(pub N);
```

Comment:
As all the other Attribute selectors are `Attr*`, what do you think about naming this `AttrExists`? or is there a precedent for using `HasAttr` name here (from other libraries)?

Reply:
I used it because the name came up in Python, but I'm fine with either; `AttrExists` might be nicer for the docs list, so, I'll change to that.